Hupper fails to reload app cleanly on Docker Compose environments

Hello all, first off thanks for hupper 🌟 

The warehouse team uses hupper in our development stack but [recently](https://github.com/pypa/warehouse/issues/6578) we started noticing the it failed to restart on code changes. I believe this is because we invoke hupper in a Docker Compose environment causes the `input` call to fail with an EOF.

We've worked around it by pinning the hupper version to 1.7 but ideally there'd be a CLI option to prevent reading from stdin. Would that be acceptable?

Sure, I guess I jumped the gun suggesting a solution rather than explaining the problem.

The symptom is that hupper tries to restart the application but it fails because the previous app did not exit and is still bound to the port. Here's a copy of the Docker logs:

```
web_1            | [2019-09-04 09:31:05 +0000] [16] [INFO] Worker exiting (pid: 16)
web_1            | Server did not exit, forcefully killing.
web_1            | Starting monitor for PID 17.
web_1            | {"logger": "raven.base.Client", "level": "INFO", "event": "Raven is not configured (logging is disabled). Please see the documentation for more information.", "thread": 140079147017984}
web_1            | [2019-09-04 09:31:31 +0000] [17] [INFO] Starting gunicorn 19.9.0
web_1            | [2019-09-04 09:31:31 +0000] [17] [ERROR] Connection in use: ('0.0.0.0', 8000)
web_1            | [2019-09-04 09:31:31 +0000] [17] [ERROR] Retrying in 1 second.
web_1            | [2019-09-04 09:31:32 +0000] [17] [ERROR] Connection in use: ('0.0.0.0', 8000)
web_1            | [2019-09-04 09:31:32 +0000] [17] [ERROR] Retrying in 1 second.
web_1            | [2019-09-04 09:31:33 +0000] [17] [ERROR] Connection in use: ('0.0.0.0', 8000)
web_1            | [2019-09-04 09:31:33 +0000] [17] [ERROR] Retrying in 1 second.
web_1            | [2019-09-04 09:31:34 +0000] [17] [ERROR] Connection in use: ('0.0.0.0', 8000)
web_1            | [2019-09-04 09:31:34 +0000] [17] [ERROR] Retrying in 1 second.
web_1            | [2019-09-04 09:31:35 +0000] [17] [ERROR] Connection in use: ('0.0.0.0', 8000)
web_1            | [2019-09-04 09:31:35 +0000] [17] [ERROR] Retrying in 1 second.
web_1            | [2019-09-04 09:31:36 +0000] [17] [ERROR] Can't connect to ('0.0.0.0', 8000)
web_1            | Press ENTER or change a file to reload.
web_1            | Traceback (most recent call last):
web_1            |   File "<string>", line 1, in <module>
web_1            |   File "/opt/warehouse/lib/python3.7/site-packages/hupper/ipc.py", line 318, in spawn_main
web_1            |     func(**kwargs)
web_1            |   File "/opt/warehouse/lib/python3.7/site-packages/hupper/worker.py", line 263, in worker_main
web_1            |     func(*spec_args, **spec_kwargs)
web_1            |   File "/opt/warehouse/lib/python3.7/site-packages/hupper/reloader.py", line 285, in wait_main
web_1            |     input('')
web_1            | EOFError: EOF when reading a line
web_1            | Starting monitor for PID 30.

... repeats endlessly...
```

The specific hupper invocation we use is in the `docker-compose.yml`: https://github.com/pypa/warehouse/blob/master/docker-compose.yml#L48

--- a/CONTRIBUTORS.txt
+++ b/CONTRIBUTORS.txt
@@ -109,3 +109,4 @@ Contributors
 - Bert JW Regeer (2017-05-17)
 - Jens Carl (2017-05-22)
 - Eric Atkin (2019-02-15)
+- Yeray Díaz Díaz (2019-10-03)

--- a/src/hupper/cli.py
+++ b/src/hupper/cli.py
@@ -6,6 +6,18 @@ from .logger import LogLevel
 from .reloader import start_reloader
 
 
+def shutdown_interval_parser(string):
+    """Parses the shutdown interval onto a float > 0.0."""
+    msg = "Shutdown interval must be a float greater than 0"
+    try:
+        value = float(string)
+        if value <= 0.0:
+            raise argparse.ArgumentTypeError(msg)
+        return value
+    except ValueError:
+        raise argparse.ArgumentTypeError(msg)
+
+
 def main():
     parser = argparse.ArgumentParser()
     parser.add_argument("-m", dest="module", required=True)
@@ -13,6 +25,9 @@ def main():
     parser.add_argument("-x", dest="ignore", action="append")
     parser.add_argument("-v", dest="verbose", action='store_true')
     parser.add_argument("-q", dest="quiet", action='store_true')
+    parser.add_argument(
+        "-i", "--shutdown-interval", type=shutdown_interval_parser
+    )
 
     args, unknown_args = parser.parse_known_args()
 
@@ -26,7 +41,10 @@ def main():
         level = LogLevel.INFO
 
     reloader = start_reloader(
-        "hupper.cli.main", verbose=level, ignore_files=args.ignore
+        "hupper.cli.main",
+        verbose=level,
+        ignore_files=args.ignore,
+        shutdown_interval=args.shutdown_interval,
     )
 
     sys.argv[1:] = unknown_args

--- a/src/hupper/cli.py
+++ b/src/hupper/cli.py
@@ -25,9 +25,7 @@ def main():
     parser.add_argument("-x", dest="ignore", action="append")
     parser.add_argument("-v", dest="verbose", action='store_true')
     parser.add_argument("-q", dest="quiet", action='store_true')
-    parser.add_argument(
-        "-i", "--shutdown-interval", type=shutdown_interval_parser
-    )
+    parser.add_argument("--shutdown-interval", type=shutdown_interval_parser)
 
     args, unknown_args = parser.parse_known_args()
 

--- a/src/hupper/cli.py
+++ b/src/hupper/cli.py
@@ -7,11 +7,11 @@ from .reloader import start_reloader
 
 
 def shutdown_interval_parser(string):
-    """Parses the shutdown interval onto a float > 0.0."""
-    msg = "Shutdown interval must be a float greater than 0"
+    """Parses the shutdown interval onto an int greater than 0."""
+    msg = "Shutdown interval must be an int greater than 0"
     try:
-        value = float(string)
-        if value <= 0.0:
+        value = int(string)
+        if value <= 0:
             raise argparse.ArgumentTypeError(msg)
         return value
     except ValueError:

--- a/tests/test_cli.py
+++ b/tests/test_cli.py
@@ -1,0 +1,15 @@
+import argparse
+
+import pytest
+
+from hupper.cli import shutdown_interval_parser
+
+
+@pytest.mark.parametrize('value', ['0.0', "-1.0"])
+def test_shutdown_interval_parser_errors(value):
+    with pytest.raises(argparse.ArgumentTypeError):
+        shutdown_interval_parser(value)
+
+
+def test_shutdown_interval_parser():
+    assert shutdown_interval_parser("5.8") == 5.8

--- a/tests/test_cli.py
+++ b/tests/test_cli.py
@@ -5,11 +5,11 @@ import pytest
 from hupper.cli import shutdown_interval_parser
 
 
-@pytest.mark.parametrize('value', ['0.0', "-1.0"])
+@pytest.mark.parametrize('value', ['0', "-1"])
 def test_shutdown_interval_parser_errors(value):
     with pytest.raises(argparse.ArgumentTypeError):
         shutdown_interval_parser(value)
 
 
 def test_shutdown_interval_parser():
-    assert shutdown_interval_parser("5.8") == 5.8
+    assert shutdown_interval_parser("5") == 5

